File: ts-model-api/src/INodeJS.ts
import type { IConceptJS } from "./IConceptJS.js";

export type ChildRole = object;
export type ReferenceRole = object;
export type PropertyRole = object;

export function toRoleJS<T extends string | undefined | null>(
  str: T
):
  | (ChildRole & ReferenceRole & PropertyRole)
  | (T extends undefined ? undefined : never)
  | (T extends null ? null : never) {
  return str as unknown as ChildRole & ReferenceRole & PropertyRole;
}

export interface INodeJS {
  getConcept(): IConceptJS | undefined;
  getConceptUID(): string | undefined;

  getReference(): INodeReferenceJS;
  getRoleInParent(): ChildRole | undefined;
  getParent(): INodeJS | undefined;

  remove(): void;

  getChildren(role: ChildRole | undefined): Array<INodeJS>;
  getAllChildren(): Array<INodeJS>;
  moveChild(role: ChildRole | undefined, index: number, child: INodeJS): void;
  addNewChild(
    role: ChildRole | undefined,
    index: number,
    concept: IConceptJS | undefined
  ): INodeJS;
  removeChild(child: INodeJS): void;

  getReferenceRoles(): Array<ReferenceRole>;
  getReferenceTargetNode(role: ReferenceRole): INodeJS | undefined;
  getReferenceTargetRef(role: ReferenceRole): INodeReferenceJS | undefined;
  setReferenceTargetNode(
    role: ReferenceRole,
    target: INodeJS | undefined
  ): void;
  setReferenceTargetRef(
    role: ReferenceRole,
    target: INodeReferenceJS | undefined
  ): void;

  getPropertyRoles(): Array<PropertyRole>;
  getPropertyValue(role: PropertyRole): string | undefined;
  setPropertyValue(role: PropertyRole, value: string | undefined): void;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Keep for backward compatibility
type INodeReferenceJS = any;

File: model-api/src/jsMain/kotlin/INodeJS.kt
import org.modelix.model.api.INode

/*
 * Copyright (c) 2023.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@JsExport
interface INodeJS {
    fun unwrap(): INode
    fun getConcept(): IConceptJS?
    fun getConceptUID(): String?
    fun getReference(): INodeReferenceJS
    fun getRoleInParent(): String?
    fun getParent(): INodeJS?
    fun getChildren(role: String?): Array<INodeJS>
    fun getAllChildren(): Array<INodeJS>
    fun moveChild(role: String?, index: Int, child: INodeJS)
    fun addNewChild(role: String?, index: Int, concept: IConceptJS?): INodeJS
    fun removeChild(child: INodeJS)
    fun getReferenceRoles(): Array<String>
    fun getReferenceTargetNode(role: String): INodeJS?
    fun getReferenceTargetRef(role: String): INodeReferenceJS?
    fun setReferenceTargetNode(role: String, target: INodeJS?)
    fun setReferenceTargetRef(role: String, target: INodeReferenceJS?)
    fun getPropertyRoles(): Array<String>
    fun getPropertyValue(role: String): String?
    fun setPropertyValue(role: String, value: String?)
}
